### ISSUE TITLE
Crash at the end of process

### ISSUE THREAD

**Issue by CharlesRA** [node-webrtc/node-webrtc#636]
Hello
When i try this minimal example on wrtc:0.45 with node 13:14 i got a segfault at the end of the process.
```
const {RTCAudioSource, RTCAudioSink} = require("wrtc").nonstandard;
const source = new RTCAudioSource();
const track = source.createTrack();
setTimeout(() => {
	track.stop();
}, 1000);
```
I got the following error 
```
#0  0x0000000004475010 in ?? ()
#1  0x00000000009e427f in node_napi_env__::~node_napi_env__() ()
#2  0x00000000009be02a in node::Environment::RunCleanup() ()
#3  0x0000000000a4e6da in node::NodeMainInstance::Run() ()
#4  0x00000000009de4a1 in node::Start(int, char**) ()
#5  0x00007ffff7a611e3 in __libc_start_main (main=0x97ba30 <main>, argc=4, argv=0x7fffffffdef8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffdee8)
    at ../csu/libc-start.c:308
#6  0x000000000097cfa5 in _start ()
```
Did I forget to do something?

**Comment by CharlesRA** [comment]
It's seem to crash with all node versions > 10

**Comment by jlitzingerdev** [comment]
I initially missed this and opened a new issue, but I'm seeing also segfault, but I've limited my testing to node versions v12.13.0, v12.13.1, v12.14.0, and 14.  v12.13.1 is the first time I see the crash (with a lot of runtime on v12.13.0).  This is on macOS 10.15.4. I've crafted a minimal reproducer:

repro.js
```
const wrtc = require('wrtc');
const sf = require('segfault-handler');

sf.registerHandler('crash.log');
wrtc.getUserMedia({ video: true }).then(() => { console.log('done'); }).catch((err) => { console.error(err); });
```
package.json
```
{
  "name": "wrtc-debug",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "segfault-handler": "^1.3.0",
    "wrtc": "0.4.6"
  }
}
```
Run with node repro.js

Anyone else seeing this? It is only happening at exit.

**Comment by jlitzingerdev** [comment]
I bisected NodeJS and found: https://github.com/nodejs/node/commit/aac2476346 to be the offending commit.

**Comment by jlitzingerdev** [comment]
An update....and this is gross, I know that.  However, I have found that if I prevent cleanup of the peer connection factory, the above reproducer no longer segfaults on v12.13.1.  I'm sure it is leaking at exit (valgrind would likely confirm).  Still, in the interest of information sharing:

```
diff --git a/src/methods/get_user_media.cc b/src/methods/get_user_media.cc
index 1c69c1d..9744a36 100644
--- a/src/methods/get_user_media.cc
+++ b/src/methods/get_user_media.cc
@@ -108,6 +108,7 @@ Napi::Value node_webrtc::GetUserMedia::GetUserMediaImpl(const Napi::CallbackInfo
   CONVERT_ARGS_OR_REJECT_AND_RETURN_NAPI(deferred, info, constraints, MediaStreamConstraints)

   auto factory = node_webrtc::PeerConnectionFactory::GetOrCreateDefault();
+  factory->SuppressDestruct();
   auto stream = factory->factory()->CreateLocalMediaStream(rtc::CreateRandomUuid());

   auto audio = constraints.audio.Map([](auto constraint) {
```

I was trying to minimize the operations I needed to do in order to fix the crash in the above case.  Note that I'm not saying this is a true fix, but it has an interesting effect.

Also prevents segfault on v12.14.0.

This feels a bit like it is just reverting to the behavior before finalizers were run.

Also, the crash will still happen if
```
ms = new MediaStream();
```
is used in place of `getUserMedia`.

**Comment by piranna** [comment]
Seems like https://github.com/node-webrtc/node-webrtc/issues/650 is also related to this.

**Comment by piranna** [comment]
And https://github.com/node-webrtc/node-webrtc/issues/641 too.

**Comment by sgodin** [comment]
I am also seeing this issue on process exit.  Node 10.15.2 was fine - no crash, then crash was introduced after updating to node 12.18.3.  Stack trace is the same as the first post here.  Is this being actively looked at right now?

**Comment by andrevmatos** [comment]
Minimally reproduce example, NodeJS v14.15.0: `$ node index.js`
```js
const wrtc = require('wrtc');
const pc = new wrtc.RTCPeerConnection();
setTimeout(() => pc.close(), 5000);
```

**Comment by andrevmatos** [comment]
I compiled NodeJS v14.15.1 with debug symbols enabled, and managed to extract some more information:

`gdb` on the segfault location:
```
(peer_connection.cc:4771): Session: 2359087827330281600 Old state: kStable New state: kClosed
(paced_sender.cc:177): ProcessThreadAttached 0x0
[Thread 0x7fffe5e25640 (LWP 434434) exited]
[Thread 0x7fffe5f26640 (LWP 434433) exited]
(paced_sender.cc:177): ProcessThreadAttached 0x0
(peer_connection.cc:6980): Usage signature is 1024
(peer_connection.cc:1076): Session: 2359087827330281600 is destroyed.
(webrtc_video_engine.cc:489): WebRtcVideoEngine::~WebRtcVideoEngine
(webrtc_voice_engine.cc:224): WebRtcVoiceEngine::~WebRtcVoiceEngine
[Thread 0x7ffff424b640 (LWP 434423) exited]
[Thread 0x7ffff414a640 (LWP 434424) exited]
[Thread 0x7ffff524d640 (LWP 434421) exited]
[Thread 0x7ffff4a4c640 (LWP 434422) exited]

Thread 1 "node_g" received signal SIGSEGV, Segmentation fault.
0x00005555561d45bb in v8impl::(anonymous namespace)::RefBase::Delete (reference=0x55555ae68810) at ../src/js_native_api_v8.cc:249
249           delete reference;
(gdb) list
244         reference->Unlink();
245         if ((reference->RefCount() != 0) ||
246             (reference->_finalize_callback == nullptr) ||
247             (reference->_delete_self) ||
248             (reference->_finalize_ran)) {
249           delete reference;
250         } else {
251           // defer until finalizer runs as
252           // it may alread be queued
253           reference->_delete_self = true;
(gdb) p *reference
$1 = {<v8impl::Finalizer> = {_env = 0x55555ae95fa0, _finalize_callback = 0x7fffe64002fa <Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::FinalizeCallback(napi_env__*, void*, void*)>,
    _finalize_data = 0x55555ae67d50, _finalize_hint = 0x0, _finalize_ran = false, _has_env_reference = false}, <v8impl::RefTracker> = {_vptr.RefTracker = 0x55500f834868, next_ = 0x0, prev_ = 0x0},
  _refcount = 1, _delete_self = false}
(gdb) bt
#0  0x00005555561d45bb in v8impl::(anonymous namespace)::RefBase::Delete (reference=0x55555ae68810) at ../src/js_native_api_v8.cc:249
#1  0x00005555561d46ac in v8impl::(anonymous namespace)::RefBase::Finalize (this=0x55555ae68810, is_env_teardown=true) at ../src/js_native_api_v8.cc:282
#2  0x000055555620632a in v8impl::RefTracker::FinalizeAll (list=0x55555ae95fd8) at ../src/js_native_api_v8.h:43
#3  0x0000555556206454 in napi_env__::~napi_env__ (this=0x55555ae95fa0, __in_chrg=<optimized out>) at ../src/js_native_api_v8.h:66
#4  0x000055555620a070 in node_napi_env__::~node_napi_env__ (this=0x55555ae95fa0, __in_chrg=<optimized out>) at ../src/node_api.cc:14
#5  0x000055555620a08c in node_napi_env__::~node_napi_env__ (this=0x55555ae95fa0, __in_chrg=<optimized out>) at ../src/node_api.cc:14
#6  0x00005555561e3007 in napi_env__::Unref (this=0x55555ae95fa0) at ../src/js_native_api_v8.h:77
#7  0x00005555562006e6 in operator() (__closure=0x0, arg=0x55555ae95fa0) at ../src/node_api.cc:100
#8  0x0000555556200706 in _FUN () at ../src/node_api.cc:101
#9  0x00005555561ac376 in node::Environment::RunCleanup (this=0x55555addb350) at ../src/env.cc:624
#10 0x0000555556140c40 in node::FreeEnvironment (env=0x55555addb350) at ../src/api/environment.cc:385
#11 0x00005555562cb440 in node::FunctionDeleter<node::Environment, &node::FreeEnvironment>::operator() (this=0x7fffffffc630, pointer=0x55555addb350) at ../src/util.h:624
#12 0x00005555562cb06e in std::unique_ptr<node::Environment, node::FunctionDeleter<node::Environment, &node::FreeEnvironment> >::~unique_ptr (this=0x7fffffffc630, __in_chrg=<optimized out>)
    at /usr/include/c++/10.2.0/bits/unique_ptr.h:361
#13 0x00005555562ca7be in node::NodeMainInstance::Run (this=0x7fffffffc7e0) at ../src/node_main_instance.cc:114
#14 0x00005555561f82b6 in node::Start (argc=2, argv=0x7fffffffca18) at ../src/node.cc:1101
#15 0x0000555557a96bc0 in main (argc=2, argv=0x7fffffffca18) at ../src/node_main.cc:141
```

The `valgrind --track-origins=yes --keep-stacktraces=alloc-and-free` for the script above ends with this entry:
```
==160991== Invalid read of size 8
==160991==    at 0xA7D8AD8: Napi::Env::operator napi_env__*() const (napi-inl.h:269)
==160991==    by 0xAB65659: node_webrtc::Deferrer::Queue() (deferrer.cc:26)
==160991==    by 0xAB5DADD: node_webrtc::AsyncContextReleaser::Release(Napi::AsyncContext*) (async_context_releaser.cc:16)
==160991==    by 0xAA9C207: node_webrtc::AsyncObjectWrap<node_webrtc::RTCPeerConnection>::DestroyAsyncContext() (async_object_wrap.h:21)
==160991==    by 0xAA9778F: node_webrtc::AsyncObjectWrap<node_webrtc::RTCPeerConnection>::~AsyncObjectWrap() (async_object_wrap.h:37)
==160991==    by 0xAA9502A: node_webrtc::AsyncObjectWrapWithLoop<node_webrtc::RTCPeerConnection>::~AsyncObjectWrapWithLoop() (async_object_wrap_with_loop.h:14)
==160991==    by 0xAA70834: node_webrtc::RTCPeerConnection::~RTCPeerConnection() (rtc_peer_connection.cc:100)
==160991==    by 0xAA70869: node_webrtc::RTCPeerConnection::~RTCPeerConnection() (rtc_peer_connection.cc:109)
==160991==    by 0xAAAEF5D: Napi::ObjectWrap<node_webrtc::RTCPeerConnection>::FinalizeCallback(napi_env__*, void*, void*) (napi-inl.h:3407)
==160991==    by 0xDBA566: napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::operator()(napi_env__*) const (js_native_api_v8.h:107)
==160991==    by 0xDBAFB2: void napi_env__::CallIntoModule<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}, void (napi_env__*, v8::Local<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::Value>)>(napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}&&, void (&&)(napi_env__*, v8::Local<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::Value>)) (js_native_api_v8.h:95)
==160991==    by 0xDBA5D9: napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*) (js_native_api_v8.h:106)
==160991==  Address 0x952ca28 is 8 bytes inside a block of size 176 free'd
==160991==    at 0x60FF08B: operator delete(void*, unsigned long) (vg_replace_malloc.c:593)
==160991==    by 0xAB5E470: node_webrtc::AsyncContextReleaser::~AsyncContextReleaser() (async_context_releaser.h:12)
==160991==    by 0xAB5E4A7: Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::FinalizeCallback(napi_env__*, void*, void*) (napi-inl.h:3407)
==160991==    by 0xDBA566: napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::operator()(napi_env__*) const (js_native_api_v8.h:107)
==160991==    by 0xDBAFB2: void napi_env__::CallIntoModule<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}, void (napi_env__*, v8::Local<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::Value>)>(napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}&&, void (&&)(napi_env__*, v8::Local<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::Value>)) (js_native_api_v8.h:95)
==160991==    by 0xDBA5D9: napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*) (js_native_api_v8.h:106)
==160991==    by 0xD8868D: v8impl::(anonymous namespace)::RefBase::Finalize(bool) (js_native_api_v8.cc:275)
==160991==    by 0xDBA329: v8impl::RefTracker::FinalizeAll(v8impl::RefTracker*) (js_native_api_v8.h:43)
==160991==    by 0xDBA453: napi_env__::~napi_env__() (js_native_api_v8.h:66)
==160991==    by 0xDBE06F: node_napi_env__::~node_napi_env__() (node_api.cc:14)
==160991==    by 0xDBE08B: node_napi_env__::~node_napi_env__() (node_api.cc:14)
==160991==    by 0xD97006: napi_env__::Unref() (js_native_api_v8.h:77)
==160991==  Block was alloc'd at
==160991==    at 0x60FDDEF: operator new(unsigned long) (vg_replace_malloc.c:342)
==160991==    by 0xAB61BBF: Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::ConstructorCallbackWrapper(napi_env__*, napi_callback_info__*)::{lambda()#1}::operator()() const (napi-inl.h:3281)
==160991==    by 0xAB64EA2: napi_value__* Napi::details::WrapCallback<Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::ConstructorCallbackWrapper(napi_env__*, napi_callback_info__*)::{lambda()#1}>(Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::ConstructorCallbackWrapper(napi_env__*, napi_callback_info__*)::{lambda()#1}) (napi-inl.h:61)
==160991==    by 0xAB61CE7: Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::ConstructorCallbackWrapper(napi_env__*, napi_callback_info__*) (napi-inl.h:3279)
==160991==    by 0xD9613B: v8impl::(anonymous namespace)::CallbackWrapperBase<v8::FunctionCallbackInfo<v8::Value>, &v8impl::(anonymous namespace)::CallbackBundle::function_or_getter>::InvokeCallback()::{lambda(napi_env__*)#1}::operator()(napi_env__*) const (js_native_api_v8.cc:479)
==160991==    by 0xD96AA2: void napi_env__::CallIntoModule<v8impl::(anonymous namespace)::CallbackWrapperBase<v8::FunctionCallbackInfo<v8::Value>, &v8impl::(anonymous namespace)::CallbackBundle::function_or_getter>::InvokeCallback()::{lambda(napi_env__*)#1}, void (napi_env__*, v8::Local<v8::Value>)>(v8impl::(anonymous namespace)::CallbackWrapperBase<v8::FunctionCallbackInfo<v8::Value>, &v8impl::(anonymous namespace)::CallbackBundle::function_or_getter>::InvokeCallback()::{lambda(napi_env__*)#1}&&, void (&&)(napi_env__*, v8::Local<v8::Value>)) (js_native_api_v8.h:95)
==160991==    by 0xD961C0: v8impl::(anonymous namespace)::CallbackWrapperBase<v8::FunctionCallbackInfo<v8::Value>, &v8impl::(anonymous namespace)::CallbackBundle::function_or_getter>::InvokeCallback() (js_native_api_v8.cc:478)
==160991==    by 0xD88D0B: v8impl::(anonymous namespace)::FunctionCallbackWrapper::Invoke(v8::FunctionCallbackInfo<v8::Value> const&) (js_native_api_v8.cc:497)
==160991==    by 0x1150C4B: v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) (api-arguments-inl.h:158)
==160991==    by 0x1154B0B: v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<true>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) (builtins-api.cc:111)
==160991==    by 0x1157368: v8::internal::Builtins::InvokeApiFunction(v8::internal::Isolate*, bool, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*, v8::internal::Handle<v8::internal::HeapObject>) (builtins-api.cc:238)
==160991==    by 0x133BD7C: v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) (execution.cc:276)
```

It seems the segfault is caused by some double-free (or free on Finalizer after it had already been cleared) of the `AsyncContextReleaser` parent of `RTCPeerConnection`. Funny thing is that it doesn't trigger if I don't `close` the connection. I tried some things, including `SuppressDestruct` on several locations as per comment above, but didn't manage to prevent the segfault. We worked around it in our project by force-exiting the process when we're done, since it only triggered on "normal teardown" of the event-loop, but of course this isn't the ideal solution. I'm afraid I can't continue looking into it.

**Comment by jlitzingerdev** [comment]
@andrevmatos We actually ended up doing the same (force exiting).  I'm hoping I can get back to it at some point but haven't been able to yet.

**Comment by netizen-ais** [comment]
Any workaround for this? I'm forced to LTS Node.js

**Comment by jlitzingerdev** [comment]
I've had to return to this and have learned a bit more.  In particular, [this](https://github.com/nodejs/node/commit/aac2476346) change upstream now triggers new behavior that previously didn't cause an issue.  The core of the problem is refcounts.  In particular, this [nodejs](https://github.com/nodejs/node/blob/e3e0927bb93ed92bcdfe81e7ad9af3d78ccc74fb/src/js_native_api_v8.cc#L224) code expects that it is safe to delete a value with a refcount > 0 when a reference is deleted and finalized.  The current node-webrtc code has a few things which have refcounts of 1 or greater.  PeerConnectionFactory is one of those things.  My SuppressDestruct suggestion above "worked" only in that avoided the ref deletion and ultimate finalization call chain.

So where does that leave it?

I think that ultimately there are expectations w.r.t object lifecycle and un-refing that aren't being met.  The didn't matter previously because these finalizers and reference destruction paths weren't run.  For example, Unref'ing PeerConnectionFactory in its destructor will prevent _one_ crash, but that is not a solution.

I'll share more if I have it, but anyone with thoughts would be most welcome.

Edit 1: 2/5/21
I'm not going to make new posts, I'm just going to edit this one.  Adding the following (which is unpleasant) to PeerConnectionFactory's destructor, resolves the first layer of issues entirely for me
```
while (_default->Unref() != 0) {}
```
I'm reasonably certain that the move-assignment operator in MediaStream::Impl needs to do an unref, which it doesn't because the factory pointer is null'd out.  The problem is that the constructors are going to take another ref.

So it's clear, I'm not advocating this as a solution, just sharing findings.  This would clearly cause issues with multiple PCF instances.

**Comment by andrevmatos** [comment]
Workaround for us, which works great because it also allows pending promises, timers and handlers to complete gracefully (or not) before exiting:
```js
process.on('beforeExit', (code) => process.exit(code));
```

Notice `beforeExit` is fired only after all pending handlers complete, and no explicit `process.exit` call took place. An explicit exit call won't trigger it. In either case, ensuring `process.exit` is called at some point (even on the `beforeExit` hook as above) is enough to prevent the wrtc segfault.

**Comment by j1elo** [comment]
Hi, I've been looking into a crash such as the one from issue https://github.com/node-webrtc/node-webrtc/issues/641, and took the chance to look closely at the node-webrtc code using this other issue as a starting point. Here is what I learned. (yeah I know this project is sadly unmaintained / abandoned by their creators, but there aren't many mature alternatives that provide with the RTCAudioSink feature like this one has, so I'm still using it for a couple things)

The crash (reproducible with https://github.com/node-webrtc/node-webrtc/issues/636#issuecomment-728574269) comes from **a use-after-free on the singleton instance AsyncContextReleaser**.

@andrevmatos showed a Valgrind trace that points to where the crash happens. I have removed all non-essential lines, and left the most relevant ones:

> The `valgrind --track-origins=yes --keep-stacktraces=alloc-and-free` for the script above ends with this entry:
>
> ```
> ==160991== Invalid read of size 8
> ==160991==    at [...]
> ==160991==    by 0xAB65659: node_webrtc::Deferrer::Queue() (deferrer.cc:26)
> ==160991==    by 0xAB5DADD: node_webrtc::AsyncContextReleaser::Release(Napi::AsyncContext*) (async_context_releaser.cc:16)
> ==160991==    by 0xAA9C207: node_webrtc::AsyncObjectWrap<...>::DestroyAsyncContext() (async_object_wrap.h:21)
> [...]
> ==160991==    by 0xAAAEF5D: Napi::ObjectWrap<node_webrtc::RTCPeerConnection>::FinalizeCallback(napi_env__*, void*, void*) (napi-inl.h:3407)
> [...]
> ==160991==  Address 0x952ca28 is 8 bytes inside a block of size 176 free'd
> ==160991==    at 0x60FF08B: operator delete(void*, unsigned long) (vg_replace_malloc.c:593)
> ==160991==    by 0xAB5E470: node_webrtc::AsyncContextReleaser::~AsyncContextReleaser() (async_context_releaser.h:12)
> ==160991==    by 0xAB5E4A7: Napi::ObjectWrap<node_webrtc::AsyncContextReleaser>::FinalizeCallback(napi_env__*, void*, void*) (napi-inl.h:3407)
> [...]
> ==160991==    by 0xDBA329: v8impl::RefTracker::FinalizeAll(v8impl::RefTracker*) (js_native_api_v8.h:43)
> [...]
> ```

So here is the thing: When Node's `RefTracker::FinalizeAll` is called, all bets are off and every JS object will get freed. That's what happened in the second half of the Valgrind  trace: `RefTracker::FinalizeAll` called `Napi::ObjectWrap::FinalizeCallback`, and this method of ObjectWrap is very clear on its intentions, [see the code here](https://github.com/nodejs/node-addon-api/blob/v5.0.0/napi-inl.h#L4593-L4598):

```cpp
instance->Finalize(Napi::Env(env));
delete instance;
```

The object instance has the opportunity to do some housekeeping *if* it overrides the virtual method `Finalize`... but after that, it gets deleted no matter what. Most importantly, at this point, the `Reference::SuppressDestruct()` does not apply at all; `delete` gets called directly. That's why attempts by @jlitzingerdev and @andrevmatos of using *SuppressDestruct* didn't really work as they intended.

And now to the crash: All *AsyncObjectWrap* objects contain a private *AsyncContext*. When the time comes and *FinalizeCallback* code runs for any *AsyncObjectWrap* object (be it an RTCPeerConnection, or really any other), its class destructor doesn't free the async context directly. Instead, it enqueues its destruction through use of a singleton instance of an *AsyncContextReleaser*. But the *AsyncContextReleaser* had already been freed by Node! And thus the **use-after-free** and crash.

**Why the AsyncContextReleaser?**

It's the very first time I touch Node native code at all, so probably I lack a lot of knowledge and context about why stuff is done the way it's done. For now, I'm at a total loss of why the *AsyncContextReleaser* class is needed, and is used in this way. But I have 2 observations:

* The *AsyncContextReleaser* class itself uses another one, *Deferrer*. This *Deferrer* class seems to reimplement in a slighltly different way what already exists in Node's [AsyncWorker](https://github.com/nodejs/node-addon-api/blob/v5.0.0/doc/async_worker.md). See the [Deferrer source](https://github.com/node-webrtc/node-webrtc/blob/v0.4.7/src/node/deferrer.cc) and [AsyncWorker source](https://github.com/nodejs/node-addon-api/blob/v5.0.0/napi-inl.h#L4794); it feels like *Deferrer* could just have been an *AsyncWorker* with either `SuppressDestruct()` or just the `OnWorkComplete()` method overriden. Which brings me to the second point...

* *Deferrer* uses [napi_create_async_work](https://nodejs.org/api/n-api.html#napi_create_async_work) with `execute = &Deferrer::DoNothing` and `complete = &Deferrer::CallExecute`. The actual deletion of our *AsyncContext* is thus done in the `complete` callback.

    This is where a bit of historical context would help me understand, because at first sight it doesn't make much sense: the point of async work is to do stuff asynchronously in the `execute` callback, not in the `complete` one... Docs state this:

    > * execute: The native function which should be called to execute the logic asynchronously. The given function is called from a worker pool thread and can execute in parallel with the main event loop thread.
    > * complete: The native function which will be called when the asynchronous logic is completed or is cancelled. **The given function is called from the main event loop thread**.

**In summary**: *AsyncContextReleaser* and *Deferrer* are used for a very roundabout way of deleting an *AsyncContext* in the main thread. But when `AsyncContextReleaser::Release` is called, the object itself might have been destroyed already.

Knowing this, I did an extremely simple thing: replaced all this machinery with a plain `delete`:

```diff
diff --git a/src/node/async_object_wrap.h b/src/node/async_object_wrap.h
index ffd2c39..2a09823 100644
--- a/src/node/async_object_wrap.h
+++ b/src/node/async_object_wrap.h
@@ -18,7 +18,7 @@ class AsyncObjectWrap: public Napi::ObjectWrap<T> {
   void DestroyAsyncContext() {
     _async_context_mutex.lock();
     if (_async_context) {
-      AsyncContextReleaser::GetDefault()->Release(_async_context);
+      delete _async_context;
       _async_context = nullptr;
     }
     _async_context_mutex.unlock();
```

Of course, this feels wrong; the releaser stuff was there for a reason. But the fact is that the crash stopped happening. Of course, no use-after-free is happening now because the *AsyncContextReleaser* is not being used at all. Valgrind trace is now clean, too. And, not sure if just by coincidence or not, but my issue with `RTCPeerConnection::OnTrack` (same as https://github.com/node-webrtc/node-webrtc/issues/641) didn't happen again so far.

Unit tests (`npm test`) don't seem to break, either (I mean not more than they already are, they don't pass as-is in the develop branch, nor in the `v0.4.7` tag, so I had to fix several of them)
